File: agent/crates/fleet-client/src/heartbeat.rs
use crate::types::{HeartbeatRequest, HeartbeatResponse};
use anyhow::Result;
use std::time::Duration;
use tokio::time;
use tonic::{transport::Channel, Request, client::Grpc, codec::ProstCodec, metadata::MetadataValue};

pub struct HeartbeatManager;

impl HeartbeatManager {
    pub async fn start(
        channel: Channel,
        token: String,
        node_id: String,
        interval_secs: u64,
    ) -> Result<()> {
        let mut interval = time::interval(Duration::from_secs(interval_secs));
        
        tokio::spawn(async move {
            loop {
                interval.tick().await;
                tracing::debug!("Sending heartbeat for node: {}", node_id);
                
                let req_payload = HeartbeatRequest {
                    node_id: node_id.clone(),
                    status: "healthy".to_string(),
                    events_buffered: 0,
                };
                
                let mut client = Grpc::new(channel.clone());
                let path = http::uri::PathAndQuery::from_static("/edr.fleet.FleetService/Heartbeat");
                let mut req = Request::new(req_payload);
                if let Ok(meta_token) = MetadataValue::try_from(format!("Bearer {}", token)) {
                    req.metadata_mut().insert("authorization", meta_token);
                }
                
                if let Err(e) = client.unary(req, path, ProstCodec::<HeartbeatRequest, HeartbeatResponse>::default()).await {
                    tracing::warn!("Failed to send heartbeat: {}", e);
                }
            }
        });
        
        Ok(())
    }
}

File: agent/crates/fleet_client/src/heartbeat.py
import asyncio
import logging

from fleet-client.types import HeartbeatRequest


logger = logging.getLogger(__name__)


class HeartbeatManager:

    @staticmethod
    async def start(
        connection,
        token: str,
        node_id: str,
        interval_secs: int,
    ) -> None:

        async def heartbeat_loop():

            while True:
                await asyncio.sleep(interval_secs)

                logger.debug(
                    "Sending heartbeat for node %s",
                    node_id,
                )

                request = HeartbeatRequest(
                    node_id=node_id,
                    status="healthy",
                    events_buffered=0,
                )

                try:
                    await connection.send_heartbeat(
                        request,
                        token,
                    )

                except Exception as e:
                    logger.warning(
                        "Failed to send heartbeat: %s",
                        e,
                    )

        asyncio.create_task(heartbeat_loop())
